frontend/src/control/fader.rs

```diff
@@ -0,0 +1,243 @@
+use wasm_bindgen::{JsCast, JsValue};
+use web_sys::{HtmlCanvasElement, CanvasRenderingContext2d, MouseEvent};
+use yew::{html, Component, ComponentLink, Html, ShouldRender, Properties, NodeRef, Callback};
+
+use crate::component::drag_target::{DragTarget, DragEvent};
+
+const FADER_WIDTH: usize = 40;
+const FADER_HEIGHT: usize = 160;
+const FADER_HANDLE_HEIGHT: usize = 17; // always odd to account for line in the middle
+const FADER_SHAFT_HEIGHT: usize = FADER_HEIGHT - FADER_HANDLE_HEIGHT;
+const FADER_SHAFT_WIDTH: usize = 4;
+const FADER_SHAFT_OFFSET_TOP: usize = FADER_HANDLE_HEIGHT / 2;
+const FADER_NOTCH_INTERVAL: usize = 11;
+
+pub struct Fader {
+    link: ComponentLink<Self>,
+    props: FaderProps,
+    canvas: NodeRef,
+    ctx: Option<CanvasRenderingContext2d>,
+    mouse_mode: MouseMode,
+}
+
+struct DragState {
+    origin_y: i32,
+    fader_value: f32,
+}
+
+enum MouseMode {
+    Normal,
+    Hover,
+    Drag(DragState),
+}
+
+#[derive(Properties, Clone)]
+pub struct FaderProps {
+    pub value: f32,
+    pub onchange: Callback<f32>,
+}
+
+pub enum FaderMsg {
+    MouseMove(MouseEvent),
+    DragStart(DragEvent),
+    Drag(DragEvent),
+    DragEnd(DragEvent),
+}
+
+impl Fader {
+    fn fader_value(&self) -> f32 {
+        match &self.mouse_mode {
+            MouseMode::Normal | MouseMode::Hover =>
+                self.props.value,
+            MouseMode::Drag(drag_state) =>
+                drag_state.fader_value,
+        }
+    }
+
+    fn fader_handle_offset_top(&self) -> f32 {
+        FADER_SHAFT_HEIGHT as f32 * (1.0 - self.fader_value())
+    }
+
+    fn drag_event(&mut self, ev: DragEvent) -> ShouldRender {
+        let origin_y = match &self.mouse_mode {
+            MouseMode::Normal | MouseMode::Hover => {
+                let origin_y = ev.offset_y;
+
+                let handle_y = self.fader_handle_offset_top();
+                let midpoint_y = handle_y + FADER_HANDLE_HEIGHT as f32 / 2.0;
+
+                let origin_y = origin_y - midpoint_y as i32;
+
+                self.mouse_mode = MouseMode::Drag(DragState {
+                    origin_y,
+                    fader_value: self.props.value,
+                });
+
+                origin_y
+            }
+            MouseMode::Drag(drag_state) => drag_state.origin_y,
+        };
+
+        let new_fader_y = ev.offset_y - origin_y;
+
+        let position = (new_fader_y - FADER_SHAFT_OFFSET_TOP as i32) as f32
+            / FADER_SHAFT_HEIGHT as f32;
+
+        let fader_value = 1.0 - position;
+
+        let fader_value = if fader_value < 0.0 {
```

Comment:
maybe we can move that `clamp` fn to a util module and then use it here?

Reply:
Hmm good suggestion but Too Hard right now because the clamps are split across crates